rust-src/concordium_base/src/transactions.rs

```diff
@@ -0,0 +1,2733 @@
+    /// execution. The unique parameters are
+    /// - `energy` -- the amount of energy that can be used for contract
+    ///   execution. The base energy amount for transaction verification will be
+    ///   added to this cost.
+    pub fn init_contract(
+        num_sigs: u32,
+        sender: AccountAddress,
+        nonce: Nonce,
+        expiry: TransactionTime,
+        payload: InitContractPayload,
+        energy: Energy,
+    ) -> PreAccountTransaction {
+        let payload = Payload::InitContract { payload };
+        make_transaction(
+            sender,
+            nonce,
+            expiry,
+            GivenEnergy::Add { num_sigs, energy },
+            payload,
+        )
+    }
+
+    /// Update a smart contract intance, giving it the given amount of energy
+    /// for execution. The unique parameters are
+    /// - `energy` -- the amount of energy that can be used for contract
+    ///   execution. The base energy amount for transaction verification will be
+    ///   added to this cost.
+    pub fn update_contract(
+        num_sigs: u32,
+        sender: AccountAddress,
+        nonce: Nonce,
+        expiry: TransactionTime,
+        payload: UpdateContractPayload,
+        energy: Energy,
+    ) -> PreAccountTransaction {
+        let payload = Payload::Update { payload };
+        make_transaction(
+            sender,
+            nonce,
+            expiry,
+            GivenEnergy::Add { num_sigs, energy },
+            payload,
+        )
+    }
+
+    /// Configure the account as a baker. Only valid for protocol version 4 and
+    /// up.
+    pub fn configure_baker(
+        num_sigs: u32,
+        sender: AccountAddress,
+        nonce: Nonce,
+        expiry: TransactionTime,
+        payload: ConfigureBakerPayload,
+    ) -> PreAccountTransaction {
+        let energy = if payload.keys_with_proofs.is_some() {
+            cost::CONFIGURE_BAKER_WITH_KEYS
+        } else {
+            cost::CONFIGURE_BAKER_WITHOUT_KEYS
+        };
+        let payload = Payload::ConfigureBaker {
+            data: Box::new(payload),
+        };
+        make_transaction(
+            sender,
+            nonce,
+            expiry,
+            GivenEnergy::Add { num_sigs, energy },
+            payload,
+        )
+    }
+
+    /// Configure the account as a delegator. Only valid for protocol version 4
+    /// and up.
+    pub fn configure_delegation(
+        num_sigs: u32,
+        sender: AccountAddress,
+        nonce: Nonce,
+        expiry: TransactionTime,
+        payload: ConfigureDelegationPayload,
+    ) -> PreAccountTransaction {
+        let payload = Payload::ConfigureDelegation { data: payload };
+        make_transaction(
+            sender,
+            nonce,
+            expiry,
+            GivenEnergy::Add {
+                num_sigs,
+                energy: cost::CONFIGURE_DELEGATION,
+            },
+            payload,
+        )
+    }
+
+    pub enum GivenEnergy {
```

Comment:
I think this could be lifted and reused a few places.

Reply:
Also doc

Reply:
Not sure what you mean by "lifted". it is used where it makes sense I believe. 

There are two slightly differnt "GivenEnergy" types, one in `construct` and one in `send`. But they need to be different since one requires more data than the other.

Reply:
I added a doc comment.